src/cj.c

```diff
@@ -0,0 +1,126 @@
+#include "data.table.h"
+
+SEXP cj(SEXP base_list) {
+  int JJ = LENGTH(base_list), nprotect = 0;
+  SEXP out = PROTECT(allocVector(VECSXP, JJ)); nprotect++;
+  int NN = 1;
+  for (int j = 0; j < JJ; j++) NN *= LENGTH(VECTOR_ELT(base_list, j));
+  int div = NN, modulo;
+  
+  //#pragma omp parallel for num_threads(getDTthreads())
+  for (int j = 0; j < JJ; j++) {
+    SEXP this_v = VECTOR_ELT(base_list, j);
+    modulo = div;
+    div = modulo/LENGTH(VECTOR_ELT(base_list, j));
+    switch(TYPEOF(this_v)) {
+    case LGLSXP: {
+      SEXP this_col = PROTECT(allocVector(LGLSXP, NN)); nprotect++;
+      int *restrict this_v_ptr = LOGICAL(this_v);
+      int *this_col_ptr = LOGICAL(this_col);
+      if (j == 0) {
+        #pragma omp parallel for num_threads(getDTthreads())
+        for (int i = 0; i < NN; i++) {
+          this_col_ptr[i] = this_v_ptr[i / div];
```

Comment:
you iterate over all rows of the output, maybe we could do `memcpy` of blocks of rows? then iterate on blocks, this can be in parallel also. One block per thread.

Reply:
Not quite sure I follow you here, sorry

Reply:
here is R code for that
```r
library(data.table)
x = 1:3; nx = length(x)
y = 1:5; ny = length(y)
ans_y = vector("integer", nx*ny)
# pragma here
for (i in seq.int(nx)) ans_y[seq.int(ny)+(i-1L)*length(y)] = y

all.equal(CJ(x, y)$y, ans_y)
```

Reply:
ohh gotcha. just loop over y once, figure out where each element goes, and copy all at once, makes sense

Reply:
Looks trivial for 2 columns, for 3+ columns gets slightly more complicated, but I believe this is the way to go if we want to have that in C optimally
```r
CJ(1:2, 1:3, 1:4)
```
```
      V1    V2    V3
    <int> <int> <int>
 1:     1     1     1
 2:     1     1     2
 3:     1     1     3
 4:     1     1     4
 5:     1     2     1
 6:     1     2     2
 7:     1     2     3
 8:     1     2     4
```